Functions/Quantification/xASL_quant_HadamardDecoding.m

```diff
@@ -163,37 +157,50 @@
 % Calculate specific data sizes
 nDecodedTI = xQ.TimeEncodedMatrixSize-1;                   % Number of TIs is always MatrixSize -1
 
-%% #997 Here we need to use something different than xQ.NumberEchoTimes to determine the structure
 %  Time decoding should be oblivious about the readout parameters
 nDecodedVolumes = xQ.NumberEchoTimes * nDecodedTI;
 
 nEncodedVolumes = size(imEncoded, 4);
 
-% #997 Here we need to use something different than xQ.NumberEchoTimes to determine the structure
 nRepetitions = nEncodedVolumes / (xQ.TimeEncodedMatrixSize * xQ.NumberEchoTimes); % Calculating no. of acquisition repeats
 
+% Check that quantification parameters are vectors with an equal length to the number of volumes
+if length(xQ.EchoTime)>1 && length(xQ.EchoTime) ~= nEncodedVolumes
```

Comment:
Nearly done.

Did you test that this also works for a single-TE?
Ideally, we have it the same way as we have it for multi-PLD use for single-PLD: instead of only doing this for multi-TE, you do the same reordering for single-TE, but the reshaping will just be identical to what it was. If you can do this, it would be great.

Also, please equalize nomenclature, E.g., you have nVolumesEncoded ~ vectorOldOrder and nVolumesDecoded ~ vectorNewOrder right?

Just trying to make it more readable/consistent, so we can make it more generalizable.

Reply:
I like this explanation: I would at least add this also to the code. I like innerDim and outerDim (dimension probably a better name than repetition). Also, I would be fine keeping the echo times, but then explaining this using the innerDim/PLD/outerDim concept (if you agree that inner and outer dim are nice specific names).

I just find it a bit tricky to do a PLD decoding based on non-PLD parameters. If we could interpret the inner dim from the PLD information, only, this would be great. But I understand that what you're saying, there are many ways for outer dims, but only a single way for an inner dim, so using the echo time is most robust. But then I would calculate this based on the EchoTime vector in combination with PLD vector; and not use "bMultiTE" for this.

E.g., if you have dual-echo (with bold as second echo) and Hadamard and we only want to use the first echo in the quantification (== no "multi-TE quantification"), but we still want to reshape because we have this inner dim, then it needs to be able to detect this.

Reply:
OK so the proposed changes are:
1. Make comments to explain innerDim/PLD/outerDim concept.
2. Calculation from EchoTime+PLD vector doesn't make sense as this is a complementary information. What we can do is to calculate from EchoTime and PLD vector separately and check if this gives the same number and report a warning otherwise. 
3. We can skip bMultiTE in Hadamard decoding as it will always have to decode it considering the innerDim anyway. And we only use bMultiTE inside the quant-FSL to really export the echos or not. So now we remove bMultiTE and in a new issue, we make sure that quant-FSL works with bMultiTE=OFF
4. Regarding your dual-echo example - yes, this will work as decoding will always take multi-TE or other innerDim into account for Hadamard (as explained above). And we will then have to modify the quant-FSL function (as explained above).

Agree with this?

Reply:
Yes I agree with this, though don't fully understand 2, because you need to compute the innerDim/PLD/outerDim (what you now call "blocks") from the EchoTime and PLD vectors, that was your whole point I thought?

Reply:
Point 2. 
- OuterDim - we already calculate correctly. These are repetitions.
- Blocks are parts of each Hadamard repetition. Hadamard-8 has 8 blocks and hadamard-4 has 4. We don't calculate that, we have a JSON field HadamardMatrixSize = 8 - which means there are 8 blocks. 
- innerDim - We only need to calculate innerDim, which is equal to Number of TE. 

So that's why, we can get the innerDimNumber from #TE or from the number of consecutive equal PLDs. From one or from the other, not from the combination of both. 

So that's why I propose:
1,3,4 - as above.
2 - Calculate innerDim# separately from #TE and separately from #of consecutive equal PLDs at the sequence start (number of the same PLDs, not the number of unique PLDs!!). And then we compare these two calculations to report an error if they don't match.

Reply:
I see. Yes, calculating innerDim# completely separate from #TE, including explanations etc, that would be my favorite. So what you say above with point 2, please add this to the explanations.

Then I would make a warning for TE not being equal to innerDim#. Because theoretically they don't have to be the same, but practically they are.

Do you agree?

Reply:
Agree and will implement that soon!